zkml/src/layers/transformer/softmax.rs

```diff
@@ -117,61 +1411,147 @@ mod tests {
+
+            let quant_softmax = softmax.quantise(qk_scaling).unwrap();
+
+            // Obtain the quantised output
+            let quant_output = quant_softmax
+                .evaluate::<GoldilocksExt2>(&[&test_qk_quant], vec![vec![num_tokens, num_tokens]])
+                .unwrap();
+            // The result of running the quantised input as floats
+            let dequant_output = softmax
+                .evaluate::<GoldilocksExt2>(&[&test_qk_dequant], vec![vec![num_tokens, num_tokens]])
+                .unwrap();
+
+            for (q_chunk, f_chunk) in quant_output.outputs[0]
+                .get_data()
+                .chunks(num_tokens)
+                .zip(dequant_output.outputs[0].get_data().chunks(num_tokens))
+            {
+                for (&q, f) in q_chunk.iter().zip(f_chunk.iter()) {
+                    let float_q = q as f32 / OUTPUT_SCALE_FACTOR as f32;
+
+                    let quant_dequant_diff = (float_q - f).abs();
+
+                    // Make sure we are always withing 1/100 th of the actual value
+                    assert!(
+                        quant_dequant_diff < 0.01,
+                        "quant dequant diff was too large got: {}",
+                        quant_dequant_diff
+                    );
+                }
+            }
+
+            let max_error =
+                quant_softmax.quant_info.as_ref().unwrap().error_bound * OUTPUT_SCALE_FACTOR as f32;
+
+            quant_output.outputs[0]
+                .get_data()
+                .chunks(num_tokens)
+                .for_each(|chunk| {
+                    let row_sum = chunk.iter().sum::<Element>();
+
+                    let diff_from_one = (row_sum - OUTPUT_SCALE_FACTOR as Element).abs();
+
+                    assert!(diff_from_one < max_error.round() as Element);
+                });
         }
     }
 
     #[test]
     fn test_softmax_with_scale() {
-        let scale = 1.0 / 2.0;
-        let softmax = Softmax::new().with_scale(scale);
-        let input = Tensor::new(vec![2, 3], vec![0.0, 0.0, 0.0, 0.0, 0.0, 0.0]);
+        let softmax = Softmax::new_with_scale(1.0 / 2.0, 1024);
+        let input = Tensor::new(
+            vec![3, 3],
+            vec![0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
+        );
         let output = softmax
-            .evaluate::<GoldilocksExt2>(&[&input], vec![vec![2, 3]])
+            .evaluate::<GoldilocksExt2>(&[&input], vec![vec![3, 3]])
             .unwrap();
 
         assert_eq!(
             output.outputs[0].get_data(),
             vec![
-                1.0 / 6.0,
-                1.0 / 6.0,
-                1.0 / 6.0,
-                1.0 / 6.0,
-                1.0 / 6.0,
-                1.0 / 6.0
+                1.0,
+                0.0,
+                0.0,
+                0.5,
+                0.5,
+                0.0,
+                1.0 / 3.0,
+                1.0 / 3.0,
+                1.0 / 3.0,
```

Comment:
Can you justify the output ? It certainly does not look like a valid output of softmax - is it only due to quantization or ... ?

Reply:
Ah gotcha, it contains the mask, makes sense.